AX: Eager tree updates done by AXObjectCache::{onFontChange, onTextColorChange} can cause re-entrant style resolution, in turn causing a crash

twilco opened the pull request:
#### 6ce568ef2b3f3a10e6dfa355f6a457256ca59481
<pre>
AX: Eager tree updates done by AXObjectCache::{onFontChange, onTextColorChange} can cause re-entrant style resolution, in turn causing a crash
<a href="https://bugs.webkit.org/show_bug.cgi?id=294146">https://bugs.webkit.org/show_bug.cgi?id=294146</a>
<a href="https://rdar.apple.com/152742167">rdar://152742167</a>

Reviewed by Joshua Hoffman.

Prior to this commit, AXObjectCache::onFontChange and AXObjectCache::onTextColorChange eagerly updated the accessibility
tree via AXIsolatedTree::updatePropertiesForSelfAndDescendants, which is not safe because the following sequence is possible:

  1. Ensure relations are dirty, e.g. by adding a new node-only element (e.g. one with display:contents) with a relation
     (e.g. aria-describedby as seen in the real webpage where this bug reproduced).
  2. Change the font for an element
  3. Force layout (e.g. via calling offsetParent on an element), which in turn forces style resolution (Style::TreeResolver::resolve)
  4. This calls AXObjectCache::onFontChange() (or AXObjectCache::onTextColorChange for the `color` CSS property)
  5. AXObjectCache::getOrCreate(WebCore::Node&amp;) is called for our new node-only, and we create it. Creating node-only
     objects can resolve relations, which happens when this bug occurs.
  6. Something calls isIgnored() as part of the creation, in turn calling AccessibilityObject::style(), in turn trying
     to resolve style. But we&apos;re already in style resolution from step 3. Crash because we triggered re-entrant style
     resolution.

This commit fixes the issue by not eagerly updating the tree in AXObjectCache::{onFontChange, onTextColorChange}. Instead,
we post a notification, then handle that to update the tree asynchronously, at which point we know layout and style are
clean.

* LayoutTests/accessibility/dynamic-font-change-dirty-style-and-relations-expected.txt: Added.
* LayoutTests/accessibility/dynamic-font-change-dirty-style-and-relations.html: Added.
* LayoutTests/accessibility/dynamic-text-color-change-dirty-style-and-relations-expected.txt: Added.
* LayoutTests/accessibility/dynamic-text-color-change-dirty-style-and-relations.html: Added.
* LayoutTests/platform/ios/TestExpectations: Enable new tests.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onFontChange):
(WebCore::AXObjectCache::onTextColorChange):
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AXObjectCache.h:

Canonical link: <a href="https://commits.webkit.org/295960@main">https://commits.webkit.org/295960@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/86c0de7ad0d9704d2e8ac1fb0545c6fb0745bceb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106695 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26446 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16843 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111905 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/57293 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108734 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27116 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34948 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81016 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109699 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21509 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96273 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61356 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20953 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14380 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56747 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90849 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14409 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114809 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33833 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24938 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90080 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34197 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92504 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89790 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22920 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34712 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12523 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29460 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33758 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/39171 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33504 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36857 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35103 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth reviewed (approved):
lgtm

I think that's a smart fix.

I wonder if it'd make sense to do the same thing for similar functions that appear to eagerly update the tree.

For example, onInertOrVisibilityChange looks superficially similar.

twilco committed:
AX: Eager tree updates done by AXObjectCache::{onFontChange, onTextColorChange} can cause re-entrant style resolution, in turn causing a crash
https://bugs.webkit.org/show_bug.cgi?id=294146
rdar://152742167

Reviewed by Joshua Hoffman.

Prior to this commit, AXObjectCache::onFontChange and AXObjectCache::onTextColorChange eagerly updated the accessibility
tree via AXIsolatedTree::updatePropertiesForSelfAndDescendants, which is not safe because the following sequence is possible:

  1. Ensure relations are dirty, e.g. by adding a new node-only element (e.g. one with display:contents) with a relation
     (e.g. aria-describedby as seen in the real webpage where this bug reproduced).
  2. Change the font for an element
  3. Force layout (e.g. via calling offsetParent on an element), which in turn forces style resolution (Style::TreeResolver::resolve)
  4. This calls AXObjectCache::onFontChange() (or AXObjectCache::onTextColorChange for the `color` CSS property)
  5. AXObjectCache::getOrCreate(WebCore::Node&) is called for our new node-only, and we create it. Creating node-only
     objects can resolve relations, which happens when this bug occurs.
  6. Something calls isIgnored() as part of the creation, in turn calling AccessibilityObject::style(), in turn trying
     to resolve style. But we're already in style resolution from step 3. Crash because we triggered re-entrant style
     resolution.

This commit fixes the issue by not eagerly updating the tree in AXObjectCache::{onFontChange, onTextColorChange}. Instead,
we post a notification, then handle that to update the tree asynchronously, at which point we know layout and style are
clean.

* LayoutTests/accessibility/dynamic-font-change-dirty-style-and-relations-expected.txt: Added.
* LayoutTests/accessibility/dynamic-font-change-dirty-style-and-relations.html: Added.
* LayoutTests/accessibility/dynamic-text-color-change-dirty-style-and-relations-expected.txt: Added.
* LayoutTests/accessibility/dynamic-text-color-change-dirty-style-and-relations.html: Added.
* LayoutTests/platform/ios/TestExpectations: Enable new tests.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onFontChange):
(WebCore::AXObjectCache::onTextColorChange):
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AXObjectCache.h:

Canonical link: https://commits.webkit.org/295960@main
